gnosis/gp-v2-services
# New system for reporting API metrics

Fixes over-reporting of API metrics for non-200 requests.

So far, we've tried two ways of reporting warp metrics. Both had issues.

The first version used the `map` filter. It didn't report requests that didn't end with an `Ok`. This is because `map` don't handle errors, and `map_err` is not available in the public warp interface. It also didn't report requests for URLs that don't match any endpoints.

The second version used one `warp::log::custom` for each endpoint. It ended up over-reporting 404s. This is becuase of how warp filters work. Consider the following setup:

```rust
    path!("endpoint_1").with(handle_metrics_for_endpoint_1)
.or(path!("endpoint_2").with(handle_metrics_for_endpoint_2))
.or(path!("endpoint_3").with(handle_metrics_for_endpoint_3))
```

In this case we have the following class structure:

```

or ---- handle_metrics_for_endpoint_1 ---- path!("endpoint_1")
|
or ---- handle_metrics_for_endpoint_2 ---- path!("endpoint_2")
|
+------ handle_metrics_for_endpoint_3 ---- path!("endpoint_3")

```

If we request `endpoint_3`, warp will run all filters top-to-bottom. It will first descend into `path!("endpoint_1")`. Then `path!("endpoint_1")` will reject url `endpoint_3` and bail an error into `handle_metrics_for_endpoint_1`. The `handle_metrics_for_endpoint_1` will report a 404 for endpoint 1 and bail an error into the first `or`. The first `or` will then call its right branch, eventually descending into `path!("endpoint_2")`... By the time we reach `path!("endpoint_3")`, a whole bunch of 404s would've been already reported.

So, the only correct way to report a request is by wrapping the top-level filter into a `warp::log::custom`.

For this to work correctly, we'll need to sanitize paths by removing variable parameters from them (i.e. replace `/api/v1/orders/<order_id>` with `/api/v1/orders/*`). I've tried my best to come up with a solution that don't require explicitly listing all allowed URLs, but failed. Unfortunately, we can't inspect warp filters and extract all endpoints and their parameters. Therefore, we have to create an explicit list of all valid URLs, and pass it to the sanitizer. I don't like it, but I don't know how to improve it.

### Test Plan
New and existing unit tests.
Manual testing:
- check that 200s are reported correctly for multiple API endpoints,
- check that 400s and 404s are reported correctly for multiple API endpoints,
- check that requests that don't match any endpoint are reported correctly.
